app/services/course/user_invitation_service.rb

```diff
@@ -0,0 +1,116 @@
+require 'csv'
+
+# Provides a service object for inviting users into a course.
+class Course::UserInvitationService
+  # Constructor for the user invitation service object.
+  #
+  # @param [Course] current_course The course to invite users to.
+  def initialize(current_course)
+    @current_course = current_course
+    @current_instance = @current_course.instance
+  end
+
+  # Invites users to the given course.
+  #
+  # @param [Array<Hash>|File|TempFile] users Invites the given users.
+  # @return [bool]
+  def invite(users)
+    Course.transaction do
+      if users.is_a?(File) || users.is_a?(Tempfile)
+        invite_from_file(users)
+      else
+        invite_users(users.deep_dup)
+      end
+
+      @current_course.save
+    end
+  end
+
+  private
+
+  # Loads the given file, then invites the users.
+  #
+  # @param [File] file Reads the given file, in CSV format, for the name and email.
+  # @return [void]
+  def invite_from_file(file)
+    invites = []
+    CSV.foreach(file, headers: true) do |row|
```

Comment:
To confirm, I just tried it, when file format is not valid, `CSV.foreach` will give exception.